File: atcoder/abc438/e.py
N, Q = map(int, input().split())
A = [int(x) for x in input().split()]

A.insert(0, 0)
N = N + 1

nxt = [[0 for _ in range(N)] for _ in range(30)]
nxt[0] = A
for i in range(1, 30):
    for u in range(N):
        nxt[i][u] = nxt[i - 1][nxt[i - 1][u]]

cnt = [[0 for _ in range(N)] for _ in range(30)]
cnt[0] = list(range(N))
for i in range(1, 30):
    for u in range(N):
        cnt[i][u] = cnt[i - 1][u] + cnt[i - 1][nxt[i - 1][u]]

res = []
for _ in range(Q):
    t, b = map(int, input().split())
    v = 0
    for i in range(30):
        if (t >> i) & 1:
            v += cnt[i][b]
            b = nxt[i][b]
    res.append(v)

print("\n".join(map(str, res)))

File: atcoder/abc438/e.rs
#![allow(unused)]

use proconio::input;

fn main() {
    input! {
        n: usize,
        q: usize,
        mut arr: [usize; n],
        ask: [(usize, usize); q],
    }

    arr.insert(0, 0);
    let n = n + 1;

    let mut nxt = vec![vec![0; n]; 30];
    let mut sum = vec![vec![0; n]; 30];

    nxt[0] = arr;
    for u in 0..n {
        sum[0][u] = u as i64;
    }

    for i in 1..30 {
        for u in 0..n {
            let v = nxt[i - 1][u];
            nxt[i][u] = nxt[i - 1][v];
            sum[i][u] = sum[i - 1][u] + sum[i - 1][v];
        }
    }

    let mut ans = vec![];
    for &(t, b) in &ask {
        let mut v = 0;
        let mut x = b;
        for i in 0..30 {
            if (t >> i) & 1 == 1 {
                v += sum[i][x];
                x = nxt[i][x];
            }
        }
        ans.push(v);
    }

    println!("{}", join(&ans, "\n"));
}

fn join<T: ToString>(arr: &[T], sep: &str) -> String {
    arr.iter()
        .map(|x| x.to_string())
        .collect::<Vec<String>>()
        .join(sep)
}
